port/wjwwood_serial_port.cpp

```diff
@@ -0,0 +1,226 @@
+        return ESP_LOADER_ERROR_FAIL;
+    }
+
+    try {
+        setTimeout(timeout);
+        size_t result = serial_port->write(data, size);
+        if (result != size) {
+            loader_port_debug_print("Could not write requested data\n");
+            return ESP_LOADER_ERROR_TIMEOUT;
+        }
+    } catch (std::exception &e){
+        loader_port_debug_print(e.what());
+        return ESP_LOADER_ERROR_FAIL;
+    }
+
+    #ifdef SERIAL_FLASHER_DEBUG_TRACE
+        transfer_debug_print(data, size, true);
+    #endif
+
+    if(write_delay)
+        std::this_thread::sleep_for(std::chrono::microseconds(200));
+
+    return ESP_LOADER_SUCCESS;
+}
+
+
+esp_loader_error_t loader_port_read(uint8_t *data, uint16_t size, uint32_t timeout)
+{
+    if(!serial_port || !serial_port->isOpen())
+    {
+        loader_port_debug_print("loader_port_read: Port not open\n");
+        return ESP_LOADER_ERROR_FAIL;
+    }
+
+    try {
+        setTimeout(timeout);
+        size_t result = serial_port->read(data, size);
+        if (result != size) {
+            loader_port_debug_print("Could not read requested data\n");
+            return ESP_LOADER_ERROR_TIMEOUT;
+        }
+    } catch (std::exception &e){
+        loader_port_debug_print(e.what());
+        return ESP_LOADER_ERROR_FAIL;
+    }
+
+    #ifdef SERIAL_FLASHER_DEBUG_TRACE
+        transfer_debug_print(data, size, false);
+    #endif
+
+    return ESP_LOADER_SUCCESS;
+}
+
+esp_loader_error_t loader_port_wjwwood_serial_init(const loader_wjwwood_serial_config_t *config)
+{
+    serial_config = *config;
+
+    try {
+        serial_port = std::make_unique<serial::Serial>(serial_config.portName, config->baudrate, serial::Timeout::simpleTimeout(config->timeout));
+        serial_port->setRTS(false);
+        serial_port->setDTR(false);
+        serial_port->setFlowcontrol(serial::flowcontrol_hardware);
+        if ( serial_port->isOpen() == false ) {
+            loader_port_debug_print("Port open failed\n");
+            return ESP_LOADER_ERROR_FAIL;
+        }
+    } catch (std::exception &e){
+        loader_port_debug_print(e.what());
+        return ESP_LOADER_ERROR_FAIL;
+    }
+    
+    return ESP_LOADER_SUCCESS;
+}
+
+esp_loader_error_t loader_port_wjwwood_serial_deinit()
+{
+    if(!serial_port || !serial_port->isOpen())
+        return ESP_LOADER_ERROR_FAIL;
+
+    try {
+        serial_port->close();
+        serial_port.reset();
+        serial_config = {};
+    } catch (std::exception &e){
+        loader_port_debug_print(e.what());
+        return ESP_LOADER_ERROR_FAIL;
+    }
+
+    return ESP_LOADER_SUCCESS;
+}
+
+void loader_port_enter_bootloader(void)
+{
+    loader_port_reset_target();
+}
+
+void loader_port_reset_target(void)
+{
+    if(!serial_port || !serial_port->isOpen())
+    {
+        loader_port_debug_print("loader_port_reset_target: Port not open\n");
+        return;
+    }
+
+    // This might be the best we can do without GPIO
```

Comment:
The wjwwood serial library seems to have support for setting DTR/RTS
https://github.com/wjwwood/serial/blob/69e0372cf0d3796e84ce9a09aff1d74496f68720/src/serial.cc#L404-L406

Questions is whether setting them independently is fast enough. From my experience: most likely not.

I've used the serial class from Qt which has similar abstractions and still had to resort to OS specific calls to switch DTR/RTS at the same time.

Reply:
I think what he means is dtr and rts would need to switch in a single call instead of the two -> set calls listed currently.

The board I'm using currently doesn't even reset with esptool.py so hard to test

Reply:
> I think what he means is dtr and rts would need to switch in a single call instead of the two -> set calls listed currently.
> 
> The board I'm using currently doesn't even reset with esptool.py so hard to test

Yes that's what I've ment.